[DateRangeInput] Pass locale to moment for formatting, part 2

AlexLandau opened the pull request:
#### Fixes #1428 for date range inputs

This PR is closely related to #1429. We didn't notice that the date range input was a separate component from the date input. This applies the same two fixes (applying the locale at parsing time and at formatting time) to the date range input.

#### Screenshot

<img width="766" alt="screen shot 2017-08-21 at 7 47 09 pm" src="https://user-images.githubusercontent.com/5733346/29546490-a2d93e90-86a9-11e7-9a6f-c660713a1347.png">

(Note: To get the screenshot of the change, I had to add "L" to the list of formats and set the locale of the DateRangeInput. Neither of these changes are currently in the PR.)

palantirtech commented:
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @AlexLandau! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

AlexLandau commented:
Note: This PR is closely related to #1429. We didn't notice that the date range input was a separate component from the date input. This applies the same two fixes (applying the locale at parsing time and at formatting time) to the date range input.

cmslewis commented on packages/datetime/src/dateRangeInput.tsx:
```diff
@@ -790,13 +790,18 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         } else {
-            return momentDate.format((format != null) ? format : this.props.format);
+            const format = (formatOverride != null) ? formatOverride : this.props.format;
+            if (this.props.locale) {
```
I noticed this same code also appeared in #1429. Let's pull this out into a utility function in `dateUtils.ts`:

Maybe:

```tsx
/**
 * Returns a date string in the provided format localized to the provided locale.
 */
export function toLocalizedDateString(momentDate: moment.Moment, format: string, locale: string) {
    const adjustedMomentDate = (locale != null) ? momentDate.locale(locale) : momentDate;
    return adjustedMomentDate.format(format);
}
```

cmslewis commented on packages/datetime/test/dateRangeInputTests.tsx:
```diff
@@ -2303,6 +2305,11 @@ describe("<DateRangeInput>", () => {
             assertInputTextsEqual(root, START_STR, "");
         });
+
+        it("Formats locale specific format strings properly", () => {
```
ubernit: should hyphenate `locale-specific` 🤓

cmslewis commented on packages/datetime/test/dateRangeInputTests.tsx:
```diff
@@ -2303,6 +2305,11 @@ describe("<DateRangeInput>", () => {
         });
+
+        it("Formats locale specific format strings properly", () => {
+            const { root } = wrap(<DateRangeInput locale="de" format="L" value={DATE_RANGE_2}/>);
```
add space before `/>`

AlexLandau commented:
@cmslewis I've addressed your comments. (I did tweak the locale type in your suggested function to "string | undefined", since it will make things easier if you ever decide to move to strict null checks and will give extra context in the meantime.)
